#### Late Days
Zero

In [1]:
import pandas as pd

# Part 1: Load the dataset

In [2]:
numerical = pd.read_csv('chronic_kidney_disease_numerical.csv')
categorical = pd.read_csv('chronic_kidney_disease_categorical.csv')

len(numerical), len(categorical)

(420, 430)

In [3]:
numerical.head()

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN


In [4]:
categorical.head()

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


# Part 2: Analyze the Dataset

### Link
https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

### Description

This is a chronic kidney disease dataset that can help us predict chronic kidney disease in patients based on symptoms. 

The `unique_id` column refers to a patient. The `Target` which has values of either `ckd` or `notckd` is a boolean label telling us whether the patient contracted chronic kidney disease or not. The rest of the columns are the features and the complete dataset has both categorical and numerical features. The ones in the `numerical` table above are numerical features and the ones in the `categorical` table above are categorical features. 

For example, `age`, `bp`, and `bgr` from `numerical` are numerical features, and `pcc`, `htn`, and `dm` from `categorical` are categorical features. 



# Part 3: Data Preprocessing

## Part 3.1 : Drop Duplicate rows

In [5]:
# Checking for Duplicates
num_dupe = numerical["unique_id"].duplicated().sum()
cat_dupe = categorical["unique_id"].duplicated().sum()


# Drop these duplicate rows
numerical = numerical.drop_duplicates(subset=['unique_id'])
categorical = categorical.drop_duplicates(subset=['unique_id'])


print("Duplicates in Numerican, Categorical:")
print(cat_dupe, num_dupe)
print("New Lengths of Numerical, Categorical:")
print(len(numerical), len(categorical))

Duplicates in Numerican, Categorical:
30 30
New Lengths of Numerical, Categorical:
390 400


## Part 3.2: Combine two differents datasets

All `unique_id`'s are not present across both tables because `categorical` has 400 rows whereas `numercial` has 390, after removing duplicate rows. This means that `numerical` is missing 10 rows, and if we joined the two tables, there would be 10 `unique_id`'s that have catagorical features but don't have numerical features. 

In [6]:
# Merge numerical and categorical tables based on unique_id

features = pd.merge(numerical, categorical, on="unique_id") 
features

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
385,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
388,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,abnormal,present,present,no,no,no,good,no,yes,ckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [7]:
# Percentage of rows that contain atleast one missing value
missing = features[features.isnull().any(axis=1)]
len(missing)/len(features) * 100

60.76923076923077

In [8]:
features.dropna(how='any', inplace=True)
features

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
7,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
13,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,...,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
379,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
381,462892,36.0,80.0,85.0,16.0,1.1,142.0,4.1,15.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.4: Sort the dataset according to the Labels

In [9]:
# Sort the dataset according to 'Target'
features = features.sort_values(by='Target')
features.reset_index(drop=True, inplace=True)
features

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,abnormal,present,present,no,no,no,good,no,yes,ckd
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,abnormal,present,notpresent,yes,yes,yes,poor,no,no,ckd
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,abnormal,present,notpresent,yes,yes,no,poor,no,yes,ckd
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [22]:
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

df_encoded = pd.get_dummies(features, columns=categorical_columns, dtype=int)

df_encoded.columns.tolist()

['unique_id',
 'age',
 'bp',
 'bgr',
 'bu',
 'sc',
 'sod',
 'pot',
 'hemo',
 'pcv',
 'wbcc',
 'rbcc',
 'al',
 'su',
 'rbc_abnormal',
 'rbc_normal',
 'pc_abnormal',
 'pc_normal',
 'pcc_notpresent',
 'pcc_present',
 'ba_notpresent',
 'ba_present',
 'htn_no',
 'htn_yes',
 'dm_no',
 'dm_yes',
 'cad_no',
 'cad_yes',
 'appet_good',
 'appet_poor',
 'pe_no',
 'pe_yes',
 'ane_no',
 'ane_yes',
 'Target_ckd',
 'Target_notckd']

In [23]:
df_encoded.head()

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,1,0,1,0,1,0,0,1,1,0
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,0,1,0,1,1,0,1,0,1,0
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,1,0,0,1,1,0,0,1,1,0
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,0,1,0,1,0,1,0,1,1,0
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,...,1,0,1,0,1,0,1,0,1,0


### Single-Column Encoding vs Multi-Column Encoding

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: 

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [12]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

# Print the dataset

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [13]:
# Normalize the all Numerical Attributes in the dataset.

# Print the dataset

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

In [14]:
#Remove that column

# Print the dataset

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [15]:
# Export the dataframe to a new csv file

# Export the dataframe to a new json file

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: 

### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [16]:
# Code to rename all the columns in the dataset

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [17]:
# Converted SQL to Pandas code

*Describe what the above code does here*